gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/reactive/handlers/api/processor/error/AbstractFailureProcessor.java

```diff
@@ -0,0 +1,121 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.gravitee.common.http.HttpHeadersValues;
+import io.gravitee.common.http.MediaType;
+import io.gravitee.gateway.api.buffer.Buffer;
+import io.gravitee.gateway.api.http.HttpHeaderNames;
+import io.gravitee.gateway.reactive.api.ExecutionFailure;
+import io.gravitee.gateway.reactive.api.context.Request;
+import io.gravitee.gateway.reactive.api.context.RequestExecutionContext;
+import io.gravitee.gateway.reactive.api.context.Response;
+import io.gravitee.gateway.reactive.core.processor.Processor;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.reactivex.Completable;
+import io.reactivex.Maybe;
+import java.util.List;
+
+/**
+ * @author David BRASSELY (david.brassely at graviteesource.com)
+ * @author Guillaume LAMIRAND (guillaume.lamirand at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public abstract class AbstractFailureProcessor implements Processor {
+
+    public static final String ID = "simple-failure-processor";
+    /**
+     * Code for an unknown caller / application     */
+    private static final String APPLICATION_NAME_ANONYMOUS = "1";
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    protected AbstractFailureProcessor() {}
+
+    @Override
+    public String getId() {
+        return ID;
+    }
+
+    @Override
+    public Completable execute(final RequestExecutionContext executionContext) {
+        return Maybe
+            .fromCallable(
+                () -> {
+                    ExecutionFailure executionFailure = executionContext.getInternalAttribute(ATTR_INTERNAL_EXECUTION_FAILURE);
+                    if (executionFailure != null) {
+                        // If no application has been associated to the request (for example in case security chain can not be processed
+                        // correctly) set the default application to track it.
+                        if (executionContext.request().metrics().getApplication() == null) {
+                            executionContext.request().metrics().setApplication(APPLICATION_NAME_ANONYMOUS);
+                        }
+
+                        return processFailure(executionContext, executionFailure);
+                    }
+                    return null;
+                }
+            )
+            .flatMapCompletable(buffer -> executionContext.response().body(buffer));
+    }
+
+    protected Buffer processFailure(final RequestExecutionContext context, final ExecutionFailure executionFailure) {
+        final Request request = context.request();
+        final Response response = context.response();
+
+        request.metrics().setErrorKey(executionFailure.key());
+        response.status(executionFailure.statusCode());
+        response.reason(HttpResponseStatus.valueOf(executionFailure.statusCode()).reasonPhrase());
+        response.headers().set(HttpHeaderNames.CONNECTION, HttpHeadersValues.CONNECTION_CLOSE);
```

Comment:
I know that this is the current behavior on the V3 but I asking myself if we could take the Jupiter opportunity to not close the connection in case of client error ?

IMHO, we should not close the connection in case of 4xx error because it is a client error and it is not necessary to force the client to reopen a new one. In case a client send bad request with an high throughput, this makes the server consume a lot of cpu for nothing (because a new connection is recreated each time the client retry a bad request...).

Properly documented with the other changes brought by the Jupiter mode, I think it makes a lot of sense and improvements.

WDYT ?

Reply:
We could create a dedicated task to this, in order to avoid introducing any breaking change during jupiter implementation.